packages/circuit-ui/hooks/useCollapsible/useCollapsible.ts

```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright 2021, SumUp Ltd.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  useState,
+  useRef,
+  useCallback,
+  RefObject,
+  KeyboardEvent,
+  MouseEvent,
+} from 'react';
+
+import { uniqueId } from '../../util/id';
+import { useAnimation } from '../useAnimation';
+
+const DEFAULT_HEIGHT = 'auto';
+
+export type CollapsibleOptions = {
+  initialOpen?: boolean;
+  duration?: number;
+  id?: string;
+};
+
+type ButtonProps = {
+  'onClick': (event: MouseEvent | KeyboardEvent) => void;
+  'tabIndex': number;
+  'aria-controls': string;
+  'aria-expanded': 'true' | 'false';
+};
+
+type ContentProps<T> = {
+  'ref': RefObject<T>;
+  'id': string;
+  'style': {
+    overflow: 'hidden';
+    willChange: 'height';
+    opacity: 1 | 0;
+    height: string | 0;
+  };
+  'aria-hidden': null | 'true';
+};
+
+type Collapsible<T> = {
+  isOpen: boolean;
+  toggleOpen: () => void;
+  getButtonProps: (props?: {
+    onClick?: (event: MouseEvent | KeyboardEvent) => void;
+  }) => ButtonProps;
+  getContentProps: (props?: {
+    style?: Record<string, string>;
+  }) => ContentProps<T>;
+};
+
+/**
+ * A hook to build accessible and smoothly animated collapsible sections.
+ * Based on https://inclusive-components.design/collapsible-sections/
+ */
+export function useCollapsible<T extends HTMLElement = HTMLElement>({
+  initialOpen = false,
+  duration = 300,
+  id,
+}: CollapsibleOptions = {}): Collapsible<T> {
+  const detailElement = useRef<T>(null);
```

Comment:
`collapsibleElement`?

Reply:
I'll use `contentElement` to align it with `getContentProps`, okay?

Reply:
Done in e756a72.

Reply:
Sorry, yes, that's perfect!